src/objects.h

```diff
@@ -884,7 +890,13 @@ EXPORT_INLINE Obj TYPE_POSOBJ(Obj obj)
 */
 EXPORT_INLINE void SET_TYPE_POSOBJ(Obj obj, Obj val)
 {
+#ifdef HPCGAP
+    MEMBAR_WRITE();
+    // Only require a read guard
+    ((Obj *) CONST_ADDR_OBJ(obj))[0] = val;
```

Comment:
I think the path forward here is indeed to handle `T_DATOBJ` consistently with the rest rather than preserving the existing implementation. I'll look into how to best to clean this up.